backend/src/main/java/ai/giskard/ml/tunnel/OuterChannelHandler.java

```diff
@@ -1,175 +1,179 @@
-            serviceChannelsIds.remove(ctx.channel().id());
-            innerServerData = Optional.empty();
-            eventBus.post(innerServerData);
+        ChannelId channelId = ctx.channel().id();
+        log.debug("Outer channel inactive {}", channelId);
+
+        if (!channelRegistry.isServiceChannel(channelId)) {
+            channelRegistry.getInnerChannelByOuterChannelId(channelId).ifPresent(innerChannel -> {
+                try {
+                    innerChannel.close().sync();
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    log.error("Interrupted while stopping inner channel {}", innerChannel.id());
+                }
+            });
         }
     }
 
     @Override
     public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
-        log.warn("Caught exception in outer server handler", cause);
+        log.error("Caught exception in outer server handler", cause);
         Channel outerChannel = ctx.channel();
         ctx.close().addListener(future -> log.info("Service channel connection is lost: {}", outerChannel.id()));
 
-        if (innerChannelIdByOuterChannel.containsKey(outerChannel.id())) {
-            Channel innerChannel = innerChannelById.get(innerChannelIdByOuterChannel.get(outerChannel.id()));
-            if (innerChannel != null) {
-                innerChannel.close().addListener(future ->
-                    log.info("Closed inner channel {} linked to closed outer channel {}", outerChannel.id(), innerChannel.id()));
-            }
-        }
+        channelRegistry.getInnerChannelByOuterChannelId(outerChannel.id()).ifPresent(innerChannel ->
+            innerChannel.close().addListener(
+                future -> log.info("Closed inner channel {} linked to closed outer channel {}", outerChannel.id(), innerChannel.id()))
+        );
     }
 
     @Override
     public void channelRead(ChannelHandlerContext ctx, Object msg) {
         SocketChannel outerChannel = (SocketChannel) ctx.channel();
         ByteBuf in = (ByteBuf) msg;
-        while (in.readableBytes() > 0) {
-            log.debug("Outer: Writing {} bytes from {}", in.readableBytes(), outerChannel.id());
-            if (innerChannelIdByOuterChannel.containsKey(outerChannel.id())) {
-                passDataDirectly(in, outerChannel);
-            } else {
-                handleServiceChannelInput(ctx, outerChannel, in);
-            }
+        log.debug("Outer: Writing {} bytes from {}", in.readableBytes(), outerChannel.id());
+        if (channelRegistry.isDataChannel(outerChannel.id())) {
+            passDataDirectly(in, outerChannel);
+        } else {
+            handleServiceChannelInput(ctx, outerChannel, in);
         }
-
     }
 
     private void handleServiceChannelInput(ChannelHandlerContext ctx, SocketChannel outerChannel, ByteBuf in) {
```

Comment:
ctx is now unused. i think Sonar will catch other things like that but it hasn't ran on the PR yet